crazy_rl/multi_agent/jax/catch/catch.py

```diff
@@ -0,0 +1,323 @@
+
+        self.target_speed = target_speed
+
+        self._init_flying_pos = init_flying_pos
+
+        self.size = size
+
+    @override
+    def _observation_space(self, agent):
+        return spaces.Box(
+            low=np.tile(np.array([-self.size, -self.size, 0], dtype=np.float32), self.num_drones + 1),
+            high=np.tile(np.array([self.size, self.size, self.size], dtype=np.float32), self.num_drones + 1),
+            shape=(3 * (self.num_drones + 1),),  # coordinates of the drones and the target
+            dtype=np.float32,
+        )
+
+    @override
+    def _action_space(self, agent):
+        return spaces.Box(low=-1 * np.ones(3, dtype=np.float32), high=np.ones(3, dtype=np.float32), dtype=np.float32)
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_obs(self, state):
+        return jdc.replace(
+            state,
+            observations=jnp.append(
+                # each row contains the location of one agent and the location of the target
+                jnp.column_stack((state.agents_locations, jnp.tile(state.target_location, (self.num_drones, 1)))),
+                # then we add agents_locations to each row without the agent which is already in the row
+                # and make it only one dimension
+                jnp.array([jnp.delete(state.agents_locations, agent, axis=0).flatten() for agent in range(self.num_drones)]),
+                axis=1,
+            ),
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _transition_state(self, state, actions, key):
+        state = self._sanitize_action(state, actions)
+
+        # Mechanics of the target
+
+        # mean of the agent's positions
+        mean = jnp.zeros(3)
+
+        for agent in range(self.num_drones):
+            mean += state.agents_locations[agent]
+
+        mean /= self.num_drones
+
+        dist = jnp.linalg.norm(mean - state.target_location[0])
+
+        surrounded = dist <= 0.2
+
+        # if the target is out of the map, put it back in the map
+        target_location = jnp.clip(
+            state.target_location
+            # go to the opposite direction of the mean of the agents
+            + ((1 - surrounded) * (state.target_location - mean) / (dist + 0.0001) * self.target_speed)
+            # if the mean of the agents is too close to the target, move the target in a random direction,
+            # slowly because it hesitates
+            + (surrounded * random.uniform(key, (3,), minval=-1, maxval=1) * self.target_speed * 0.1),
+            jnp.array([-self.size, -self.size, 0.2]),
+            jnp.array([self.size, self.size, self.size]),
+        )
+
+        return jdc.replace(state, target_location=target_location)
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _sanitize_action(self, state, actions):
+        # Actions are clipped to stay in the map and scaled to do max 20cm in one step
+        return jdc.replace(
```

Comment:
Sorry to be boring haha. But I think `_sanitize_action` should restrict to only scale and clip the action (so it returns an array that is the sanitized action), then the agent locations can be mutated in the `_transition_state`